master/internal/audit_test.go

```diff
@@ -0,0 +1,125 @@
+package internal
+
+import (
+	"context"
+	"net/http"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/labstack/echo/v4"
+	"github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/require"
+
+	detContext "github.com/determined-ai/determined/master/internal/context"
+	"github.com/determined-ai/determined/master/pkg/model"
+)
+
+type logStore struct {
+	inner []*logrus.Entry
+}
+
+func (l *logStore) Fire(e *logrus.Entry) error {
+	l.inner = append(l.inner, e)
+	return nil
+}
+
+func (l *logStore) Levels() []logrus.Level {
+	return logrus.AllLevels
+}
+
+func TestAuditLogMiddleware(t *testing.T) {
+	// Given an echo server with our middleware, where we can introspect logs, and with a valid
+	// DetContext.
+	const url = "localhost:8081"
+
+	logs := logStore{}
+	logrus.SetLevel(logrus.DebugLevel)
+	logrus.AddHook(&logs)
+
+	e := echo.New()
+	e.Use(func(h echo.HandlerFunc) echo.HandlerFunc {
+		return func(c echo.Context) error {
+			cc := &detContext.DetContext{Context: c}
+			cc.SetUser(model.User{
+				Username: "brad",
+			})
+			return h(cc)
+		}
+	})
+	e.Use(auditLogMiddleware())
+	e.Any("/ok", echo.HandlerFunc(func(c echo.Context) error {
+		return nil
+	}))
+	e.Any("/notok", echo.HandlerFunc(func(c echo.Context) error {
+		return echo.NewHTTPError(http.StatusUnauthorized)
+	}))
+	go func() {
+		if err := e.Start(url); err != nil {
```

Comment:
I believe it'd always print an error on shutdown, but `go test` suppresses it as long as the test is ok

```suggestion
		if err := e.Start(url); err != nil && err != http.ErrServerClosed {
```

Reply:
neat, was unaware of this err type